gcc/config/loongarch/loongarch-def.h

```diff
@@ -0,0 +1,135 @@
+/* LoongArch definitions.
+   Copyright (C) 2021 Free Software Foundation, Inc.
+   Contributed by Loongson Ltd.
+
+This file is part of GCC.
+
+GCC is free software; you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation; either version 3, or (at your option)
+any later version.
+
+GCC is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with GCC; see the file COPYING3.  If not see
+<http://www.gnu.org/licenses/>.  */
+
+/* Definition of standard codes for:
+    - base architecture types  (isa_base),
+    - ISA extensions           (isa_ext),
+    - base ABI types           (abi_base),
+    - ABI extension types      (abi_ext).
+
+    - code models                    (cmodel)
+    - other command-line switches    (switch)
+
+   These values are primarily used for implementing option handling
+   logic in "loongarch.opt", "loongarch-driver.c" and "loongarch-opt.c".
+
+   As for the result of this option handling process, the following
+   scheme is adopted to represent the final configuration:
+
+    - The target ABI is encoded with a tuple (abi_base, abi_ext)
+      using the code defined below.
+
+    - The target ISA is encoded with a "struct loongarch_isa" defined
+      in loongarch-cpu.h.
+
+    - The target microarchitecture is represented with a cpu model
+      index defined in loongarch-cpu.h.
+*/
+
+#include "loongarch-tune.h"
+
+/* enum isa_base */
+extern const char* loongarch_isa_base_strings[];
+#define ISA_BASE_LA64V100     0
+#define N_ISA_BASE_TYPES      1
+
+/* enum isa_ext_* */
+extern const char* loongarch_isa_ext_strings[];
+#define ISA_EXT_NOFPU         0
+#define ISA_EXT_FPU32         1
+#define ISA_EXT_FPU64         2
+#define N_ISA_EXT_FPU_TYPES   3
+#define N_ISA_EXT_TYPES       3
+
+/* enum abi_base */
+extern const char* loongarch_abi_base_strings[];
+#define ABI_BASE_LP64D        0
+#define ABI_BASE_LP64F        1
+#define ABI_BASE_LP64S        2
+#define N_ABI_BASE_TYPES      3
+
+/* enum abi_ext */
+extern const char* loongarch_abi_ext_strings[];
+#define ABI_EXT_BASE          0
+#define N_ABI_EXT_TYPES       1
+
+/* enum cmodel  */
+extern const char* loongarch_cmodel_strings[];
+#define CMODEL_NORMAL         0
+#define CMODEL_TINY           1
+#define CMODEL_TINY_STATIC    2
+#define CMODEL_LARGE          3
+#define CMODEL_EXTREME        4
+#define N_CMODEL_TYPES        5
+
+/* enum switches */
+extern const char* loongarch_switch_strings[];
+#define SW_SOFT_FLOAT         0
+#define SW_SINGLE_FLOAT       1
+#define SW_DOUBLE_FLOAT       2
+#define N_SWITCH_TYPES        3
```

Comment:
This seems to be controlling FPU characteristics, so maybe "switch" is a bit too generic here?

Reply:
Yes. These codes represent `-m{soft,single,double}-float` switches only. They can be used to override '-mfpu=*' settings.

Reply:
Then I suggest a name like `fp_model`; for example the enum `cmodel` above is also a CLI switch, yet it doesn't have a `switch` appended to its name.

Reply:
Well, the plan is to encode any option that does not accept parameters with `SW_`. The distinction between `SW_`  and other "macro enumerations" is based upon their "syntax" rather than "semantics".

Reply:
OK, now I know better. A name like `PARAMLESS_SW_xxx` would tremendously help readability IMO; and you could certainly try to make the name shorter but that's my try.

Reply:
And your clarification comments here could be placed in the code as well, that would certainly help people understand your intention here.